File: ConfigCenterCompiler/src/main/java/com/example/configcentercompiler/Logger.java
package com.example.configcentercompiler;

/**
 * Created by 张宇 on 2018/2/5.
 * E-mail: zhangyu4@yy.com
 * YY: 909017428
 */

import javax.annotation.processing.Messager;
import javax.tools.Diagnostic;

/**
 * Simplify the messager.
 *
 * @author Alex <a href="mailto:zhilong.liu@aliyun.com">Contact me.</a>
 * @version 1.0
 * @since 16/8/22 上午11:48
 */
public class Logger {

    private static final String PREFIX_OF_LOGGER = "ConfigCenter: ";

    private final Messager msg;

    private boolean isNotEmpty(CharSequence s) {
        return s != null && s.length() != 0;
    }

    public Logger(Messager messager) {
        msg = messager;
    }

    /**
     * Print info log.
     */
    public void info(CharSequence info) {
        if (isNotEmpty(info)) {
            msg.printMessage(Diagnostic.Kind.NOTE, PREFIX_OF_LOGGER + info);
        }
    }

    public void error(CharSequence error) {
        if (isNotEmpty(error)) {
            msg.printMessage(Diagnostic.Kind.ERROR, PREFIX_OF_LOGGER + "exception:[" + error + "]");
        }
    }

    public void error(Throwable error) {
        if (null != error) {
            msg.printMessage(Diagnostic.Kind.ERROR, PREFIX_OF_LOGGER + "An exception is encountered, [" + error.getMessage() + "]" + "\n" + formatStackTrace(error.getStackTrace()));
        }
    }

    public void warning(CharSequence warning) {
        if (isNotEmpty(warning)) {
            msg.printMessage(Diagnostic.Kind.WARNING, PREFIX_OF_LOGGER + warning);
        }
    }

    private String formatStackTrace(StackTraceElement[] stackTrace) {
        StringBuilder sb = new StringBuilder();
        for (StackTraceElement element : stackTrace) {
            sb.append("    at ").append(element.toString());
            sb.append("\n");
        }
        return sb.toString();
    }
}

File: ConfigCenter/src/main/java/com/example/configcenter/Logger.kt
package com.example.configcenter

/**
 * Created by 张宇 on 2018/2/7.
 * E-mail: zhangyu4@yy.com
 * YY: 909017428
 */
internal interface Logger : ILog {
    var logger: ILog
}

internal class ConfigLogger : Logger {
    override var logger: ILog = Quiet

    override fun i(info: String) = logger.i(info)

    override fun e(error: Throwable) = logger.e(error)

    override fun d(info: String) = logger.d(info)
}

object Quiet : ILog {
    override fun i(info: String) {}

    override fun d(info: String) {}

    override fun e(error: Throwable) {}
}

interface ILog {
    fun i(info: String)

    fun d(info: String)

    fun e(error: Throwable)
}